src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/LicenceService.kt

```diff
@@ -0,0 +1,65 @@
+package uk.gov.justice.digital.hmpps.hmppshdcapi.licences
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.hmppshdcapi.model.HdcLicence
+
+@Service
+class LicenceService(
+  private val licenceRepository: LicenceRepository,
+  private val objectMapper: ObjectMapper,
+) {
+  fun getByBookingId(bookingId: Long): HdcLicence? {
+    val licenceObject = licenceRepository.findLicenceByBookingId(bookingId) ?: return null
+
+    val licence = licenceObject.licence
+
+    if (licence.isNullOrEmpty()) {
+      return null
+    }
+
+    val cas2ReferralObject = licence["bassReferral"]
+    val cas2Referral = objectMapper.convertValue(cas2ReferralObject, Cas2Referral::class.java)
+
+    val curfewObject = licence["curfew"]
+    val curfew = objectMapper.convertValue(curfewObject, Curfew::class.java)
+
+    val proposedAddressObject = licence["proposedAddress"]
+    val proposedAddress = objectMapper.convertValue(proposedAddressObject, ProposedAddress::class.java)
+
+    return transformToHdcLicence(
+      curfewAddress = getAddress(curfew, cas2Referral, proposedAddress),
+      firstNightCurfewHours = curfew.firstNight,
+      curfewHours = curfew.curfewHours,
+    )
+  }
+
+  fun getAddress(curfew: Curfew, cas2Referral: Cas2Referral, proposedAddress: ProposedAddress): CurfewAddress {
+    val isCurfewApprovedPremisesRequired = curfew.approvedPremises?.required == Decision.YES
+    val isCas2ApprovedPremisesRequired = cas2Referral.bassAreaCheck?.approvedPremisesRequiredYesNo == Decision.YES
+    val isCas2Requested = cas2Referral.bassRequest.bassRequested == Decision.YES
+    val isCas2Accepted = cas2Referral.bassOffer?.bassAccepted == OfferAccepted.YES
+
+    if (isCurfewApprovedPremisesRequired && !isCas2ApprovedPremisesRequired) {
+      return formatAddress(curfew.approvedPremisesAddress!!)
+    }
+
+    if (isCas2ApprovedPremisesRequired) {
+      return formatAddress(cas2Referral.approvedPremisesAddress!!)
```

Comment:
slightly sceptical that just because the flags are set that these addresses will always not be null.. but happy to go with it for now.

Reply:
Louise and I checked this and used the logic [here](https://github.com/ministryofjustice/licences/blob/109739c579dca1bf88bf32a866cd171be009e5c7/server/services/utils/pdfFormatter/pdfFormatter.js#L76) to come to this conclusion. Checked some data in the DB and seemed to be the case but yes, cautious optimism at best.

Reply:
yes, I think the logic for choosing which address to use is correct but think the result may be nullable - which would blow up with `!!`.